File: Graph/Shortest_path_revisited.cpp
#include <vector>
#include <cmath>
#include <algorithm>
#include <iostream>
#include <queue>

using namespace std;

struct Node{
    int city = 0;
    int cost = 0;
    int passes_left = 0;
};

class CompareDis{
    public:
        bool operator() (const pair<int, int>& n1, const pair<int, int>& n2){
            return n1.second > n2.second;
        }
};

int main(){
    
    int n, m, k;
    cin >> n >> m >> k;

    vector<vector<pair<int, int>>> G(n+1, vector<pair<int, int>>());
    
    int u, v, w;
    for(int i = 0; i < m; i++){
        cin >> u >> v >> w;
        G[u].push_back({v, w});
        G[v].push_back({u, w});
    }

    vector<vector<bool>> visited(n+1, vector<bool>(k+1, false));
    // priority_queue<Node,  vector<Node>, CompareDis> Q;
    queue<pair<int , pair<int, int>>> Q;
    vector<int> cost_to_city(n+1, INT32_MAX);

    Q.push({1, 0});

    while(!Q.empty())
    {
        pair<int, int> cur = Q.front();
        int cur_city = cur.first;
        int cost = cur.second;

        cost_to_city[cur_city] = min(cost_to_city[cur_city], cost);
        cout << cur_city << " " <<cost << endl;
        Q.pop();

        for(auto nxt: G[cur_city]){
            //cout << cur_city << " " << nxt.first << " " << passes_left << " " << cost << endl; 
            if(!visited[cur_city][nxt.first]){
                // cout << cur_city << " " << nxt.first << " " << passes_left << " " << cost << endl;
                Q.push({nxt.first, cost + nxt.second, passes_left});
                if(passes_left > 0){
                    Q.push({nxt.first, cost, passes_left-1});
                }
               
                visited[cur_city][nxt.first] = true;
            }    
        }
    }
    
    for(auto x: cost_to_city) cout << x << " ";
    
}

/*
4 5 1
1 2
2 3
1 4
3 4
2 4

*/

File: Graph/shortest_path_revisited.py

def main():
    N, M, K = map(int, input().split())

    G = [ [] for _ in range(N+1)]

    for i in range(M):
        u, v, c = map(int, input().split())
        G[u].append((v, c))
        G[v].append((u, c))

    DP = [ [float('inf')]*(K+1) for _ in range(N+1)]
    
    for i in range(K+1):
        DP[1][i] = 0

    visited = set()
    q = []
    q.append((1, 0))

    while q:
        cur_node, d = q.pop(0);
        for n, w in G[cur_node]:
            if (n, cur_node)not in visited:
                cur_d = d + 1
                DP[n][0] = min(DP[n][0], DP[cur_node][0] + w)
                for i in range(1, K+1):
                    DP[n][i] = min(DP[n][i], min(DP[cur_node][i-1], DP[cur_node][i] + w))
                q.append((n, cur_d))
                visited.add((n, cur_node))
    
    for i in DP[1:]:
        if i[k] == float('inf'):
            print(0, end = ' ')
        else:
            print(i[K], end = ' ')

if __name__ == '__main__':
    main()

# Input
# Input
"""
12 15 1
1 3 6
1 4 4
1 2 8
3 5 9
4 5 9
5 9 1
9 8 3
8 7 6
8 6 2
7 6 5
6 2 6
1 10 1
10 11 1
11 12 15
12 9 1

"""